### PR TITLE
Sampler refactoring

### PR DESCRIPTION
**Pull request by petersteneteg**
Core/Base/VecVis: Sampler refactoring. Use a single type for the sample return value

Fixes #...

Changes proposed in this PR:
 * ...
 * ...
 * ...

This has been tested on: ...

> Thank you for contributing to the Inviwo repository! Please ensure the following: 
> - Code is error and warning free
> - Code follows the Inviwo coding guidelines: 
>   https://github.com/inviwo/inviwo/wiki/Coding-Conventions
> - New Code / Processors follows the Inviwo documentation guidelines: 
>   https://github.com/inviwo/inviwo/wiki/Documentation-guide
> - New Code / processors are tested in a unittest and / or regression test: 
>   https://github.com/inviwo/inviwo/wiki/Manual#testing
> 
> Please remember to specify the environment this code been compiled and tested on:
> Operating System / Compiler / IDE / QT / CMake / Python versions.

### CODE REVIEW

File: modules/base/src/processors/volumesequencesingletimestepsampler.cpp [hunk 1]
```diff
@@ -51,24 +51,55 @@
...
 const ProcessorInfo VolumeSequenceSingleTimestepSamplerProcessor::processorInfo_{
     "org.inviwo.VolumeSequenceSingleTimestepSampler",  // Class identifier
     "Volume Sequence Single Timestep Sampler",         // Display name
     "Spatial Sampler",                                 // Category
     CodeState::Experimental,                           // Code state
     Tags::None,                                        // Tags
-};
+    R"(Creates a spatial sampler for a given timestamp from a VolumeSequence. Will use linear
+    interpolation to sample between two volume in the sequence.
+    Useful for streamline visualization of a specific time step)"_unindentHelp};
```
Comment (martinfalk) [inline modules/base/src/processors/volumesequencesingletimestepsampler.cpp]:
```suggestion
    R"(Creates a spatial sampler for a given timestamp from a VolumeSequence. Will use linear
    interpolation to sample between two adjacent volumes in the sequence.
    Useful for streamline visualization of a specific time step.)"_unindentHelp};
```

File: modules/base/src/processors/volumesequencesingletimestepsampler.cpp [hunk 2]
```diff
@@ -51,24 +51,55 @@
...
 }
 
 VolumeSequenceSingleTimestepSamplerProcessor::VolumeSequenceSingleTimestepSamplerProcessor()
     : Processor()
-    , volumeSequence_("volumeSequence")
-    , sampler_("sampler")
-    , timestamp_("timestamp", "Timestamp") {
+    , volumeSequence_("volumeSequence", "The input sequence of volumes"_help)
+    , sampler_("sampler", "The created sampler"_help)
+    , timestamp_("timestamp", "Timestamp", "the timestamp to sample at"_help) {
```
Comment (martinfalk) [inline modules/base/src/processors/volumesequencesingletimestepsampler.cpp]:
```suggestion
    , timestamp_("timestamp", "Timestamp", "The timestamp to sample at"_help) {
```

**Review by martinfalk — APPROVED** [review]
Nice work!

**Review by martinfalk — DISMISSED** [review]

File: include/inviwo/core/util/imagesampler.h [hunk 3]
```diff
@@ -47,94 +47,89 @@ class Layer;
...
+template <typename ReturnType = dvec4>
+class ImageSampler : public SpatialSampler<ReturnType> {
 public:
-    using Space = SpatialSampler<DataDims, T>::Space;
-
     /**
-     * Creates a ImageSpatialSampler for the given LayerRAM, does not take ownership of ram.
-     * Use ImageSpatialSampler(std::shared_ptr<const Image>) to ensure that the LayerRAM is
-     * available for the lifetime of the ImageSpatialSampler
+     * Creates a ImageSampler for the given LayerRAM, does not take ownership of ram.
```
Comment (ResearchDaniel) [inline include/inviwo/core/util/imagesampler.h]:
```suggestion
     * Creates an ImageSampler for the given LayerRAM. Does not take ownership of supplied data.
```

File: include/inviwo/core/util/imagesampler.h [hunk 4]
```diff
@@ -47,94 +47,89 @@ class Layer;
...
+     * Use ImageSampler(std::shared_ptr<const Image>) to ensure that the LayerRAM is
```
Comment (ResearchDaniel) [inline include/inviwo/core/util/imagesampler.h]:
```suggestion
     * @note Use ImageSampler(std::shared_ptr<const Image>) to ensure that the LayerRAM is
```

File: modules/base/include/modules/base/processors/layertospatialsampler.h [hunk 5]
```diff
@@ -48,7 +48,7 @@ class IVW_MODULE_BASE_API LayerToSpatialSampler : public Processor {
 
 private:
     LayerInport layer_;
-    DataOutport<SpatialSampler<2, double>> sampler_;
+    DataOutport<SpatialSampler<dvec4>> sampler_;
```
Comment (ResearchDaniel) [inline modules/base/include/modules/base/processors/layertospatialsampler.h]: Wrong?
```suggestion
    DataOutport<SpatialSampler<dvec2>> sampler_;
```

File: modules/base/src/processors/layertospatialsampler.cpp [hunk 6]
```diff
@@ -52,7 +52,7 @@ LayerToSpatialSampler::LayerToSpatialSampler()
 }
 
 void LayerToSpatialSampler::process() {
-    auto sampler = std::make_shared<ImageSpatialSampler<2, double>>(layer_.getData());
+    auto sampler = std::make_shared<ImageSampler<dvec4>>(layer_.getData());
```
Comment (ResearchDaniel) [inline modules/base/src/processors/layertospatialsampler.cpp]: Wrong?
```suggestion
    auto sampler = std::make_shared<ImageSampler<dvec2>>(layer_.getData());
```

File: modules/base/src/processors/volumesequencesingletimestepsampler.cpp [hunk 7]
```diff
@@ -51,24 +51,54 @@
...
+
+protected:
+    virtual dvec3 sampleDataSpace(const dvec3& pos) const override {
+        const auto a = v0_.sample(pos, CoordinateSpace::Data);
+        const auto b = v1_.sample(pos, CoordinateSpace::Data);
+        return a + t_ * (b - a);
+    };
+
+    virtual bool withinBoundsDataSpace(const dvec3& pos) const override {
+        return v0_.withinBounds(pos, CoordinateSpace::Data) ||
```
Comment (ResearchDaniel) [inline modules/base/src/processors/volumesequencesingletimestepsampler.cpp]:
```suggestion
        return v0_.withinBounds(pos, CoordinateSpace::Data) &&
```

**Review by ResearchDaniel — CHANGES_REQUESTED** [review]

File: modules/base/src/processors/layertospatialsampler.cpp (hunk 6 above)
Comment (petersteneteg) [reply modules/base/src/processors/layertospatialsampler.cpp]: Wanted to change it to dvec4 but that has to wait until a future pr.

**Review by ResearchDaniel — APPROVED** [review]

**Pull request merged**